### DISCUSSION TITLE
[help] Is there a way to keep a target `map()` pattern while running something across the full target?

### DISCUSSION THREAD

**Discussion by billdenney** [ropensci/targets#1016] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

This is tangentially related to my last question.

I have a scenario where I load many files, perform many analyses on each of the files individually, and then I summarize the results.  For each of these steps, I want to run on the mapped pattern input so that I keep my analyses on prior files when adding new ones.

But, I also need to verify that the inputs are valid (mainly that they are not duplicated in certain ways-- that a file doesn't come from the same clinical trial subject on the same date), and that comes from the contents of the files.

Question: Is there a way that I can map across a target and also do a cross-dynamic-target check to make sure that the inputs are valid?

My initial thoughts look something like the following (this is not a running code reprex, but it gives the flavor).  But, the method feels like a kludge.  Is there a better way?

```r
library(targets)
library(tarchetypes)

list(
  tar_files(
    filenames,
    list.files("datadir")
  ),
  
  tar_target(
    raw_data_prep,
    make_raw_data_prep(filenames),
    pattern = map(filenames)
  )
  
  tar_target(
    validated,
    # This function gives an informative error if there is a validation issue
    # with the raw data
    make_validation(raw_data_prep)
  ),
  
  # Ensure validation
  tar_target(
    raw_data,
    {
      # Depend on the data being validated
      validated
      # Then return the data itself
      raw_data_prep
    },
    pattern = map(raw_data_prep)
  ),
  
  tar_target(
    final_processing,
    make_final_processing(raw_data)
  )
)
```

**Comment by wlandau** [comment, marked as answer]
Sorry, I am not sure I know what you mean, and I am having trouble with the `_targets.R` file because the graph contains a cycle. 

As with my answer to #1014, I do not think it is worth creating a separate set of targets for validation, and I recommend performing the data validation checks in the same target as the analysis. Data checks are usually not computationally expensive, and the result is a simple success or failure (or a character string describing the error). The validation result can be factored into the result of the analysis, e.g. by returning `NULL` for failures if you only want to analyze the successes.

**Reply by billdenney** [reply]
Sorry for my error.  I corrected the original comment so that `make_validation(raw_data_prep)` now uses `raw_data_prep` instead of `raw_data`.

You're correct that the data checks are not computationally expensive, but this is a different use case than #1014.  I'll try to explain better now.

In #1014, I was able to validate the results of an individual file from just the data within that file.  In this, I need to look at the data from all files to validate the results.  I will describe more specifically below, and I will clarify my exact use case.

1. I receive many files that are CT scans of patients in a cancer clinical trial.  I receive the scans over time, as data are accumulated over time, and due to the way that the scans are generated (one scan at one clinical site at a time), it's not feasible for the files to be aggregated prior to my receipt.
2. Sometimes, there are errors in the scans, and revised scans are received.  I will then have two files that relate to the same person with a scan on the same date, but with two different assessment dates.  These will be mixed in with hundreds of other scan files, and the filenames do not fully clarify the person/scan date/assessment date combination.
3. My current method:
    1. Loads all of the files (with a `map(filenames)` pattern like above).
    2. Next, I want to verify that I don't have any duplicates at the person/scan date level.  I want the pipeline to stop if there are duplicates so that I can clean up the duplicate files.
    3. As part of the above step, I want to ensure that no analysis is completed unless validation is successful, even if I do a `tar_make(names = "final_analysis")` which may not directly depend on the validation step, unless forced to do so.
4. I want to keep the ability to do something like `map(raw_data_prep)` after the validation step.  The reason I want to be able to keep that is that the subsequent computations are expensive and can mostly be done at the level of a single file.  When most of the files are static, I don't want to redo all of those analyses.

Does that clarify where my question originates and what I'm trying to do?

**Reply by wlandau** [reply]
If I understand correctly, I think we can simplify the pipeline to do the validation between the data and the analyses. This would hopefully avoid having 2 copies of every dataset, and it may reduce the number of targets.

```r
# _targets.R
library(targets)
library(tarchetypes)
list(
  tar_files(files, list.files("data")),
  tar_target(data, read.csv(files), pattern = map(files)),
  tar_target(validated, is_okay(data)),
  tar_target(analysis, run_analysis(data, validated), pattern = map(data)),
  tar_target(final_processing, summarize_analyses(analysis))
)
```

**Reply by billdenney** [reply]
@wlandau, thanks.  That makes sense, and it's close to what I'm doing.
